File: src/main/java/io/github/mg138/tsbook/setting/util/ConfigBuilder.java
package io.github.mg138.tsbook.setting.util;

import org.bukkit.configuration.ConfigurationSection;
import org.bukkit.configuration.InvalidConfigurationException;
import org.bukkit.configuration.file.YamlConfiguration;
import org.bukkit.plugin.java.JavaPlugin;

import java.io.File;
import java.io.IOException;
import java.util.*;
import java.util.jar.JarEntry;
import java.util.jar.JarFile;

@SuppressWarnings("ResultOfMethodCallIgnored")
public class ConfigBuilder {
    final JavaPlugin plugin;
    final File jar;

    public ConfigBuilder(JavaPlugin plugin, File jar) {
        this.plugin = plugin;
        this.jar = jar;
    }

    public YamlConfiguration create(String path, String target) {
        File file = new File(plugin.getDataFolder(), path + target);

        if (!file.exists()) {
            file.getParentFile().mkdirs();
            plugin.saveResource(path + target, false);
        }

        YamlConfiguration yaml = new YamlConfiguration();

        try {
            yaml.load(file);
        } catch (IOException | InvalidConfigurationException e) {
            e.printStackTrace();
        }
        return yaml;
    }

    public YamlConfiguration createFolder(String path, String target) {
        List<File> files = loadJarContent(path);

        YamlConfiguration yaml = new YamlConfiguration();

        files.forEach((value) -> {
            if (value.getName().equals(target)) {
                try {
                    yaml.load(value);
                } catch (IOException | InvalidConfigurationException e) {
                    e.printStackTrace();
                }
            }
        });

        return yaml;
    }

    public List<YamlConfiguration> createFolder(String path) {
        List<File> files = loadJarContent(path);

        List<YamlConfiguration> YAMLs = new ArrayList<>();

        for (File t : files) {
            YAMLs.add(YamlConfiguration.loadConfiguration(t));
        }

        return YAMLs;
    }

    public Map<String, YamlConfiguration> createMap(String path, String key) {
        List<File> files = loadJarContent(path);

        Map<String, YamlConfiguration> MappedYAML = new HashMap<>();
        try {
            for (File file : files) {
                YamlConfiguration yaml = new YamlConfiguration();
                yaml.load(file);
                MappedYAML.put(yaml.getString(key), yaml);
            }
        } catch (IOException | InvalidConfigurationException e) {
            e.printStackTrace();
            return null;
        }

        return MappedYAML;
    }

    public Map<String, ConfigurationSection> createSectionMap(String path) {
        List<File> files = loadJarContent(path);

        Map<String, ConfigurationSection> MappedYAML = new HashMap<>();
        try {
            for (File file : files) {
                YamlConfiguration yaml = new YamlConfiguration();
                yaml.load(file);
                for (String section : yaml.getKeys(false)) {
                    MappedYAML.put(section, yaml.getConfigurationSection(section));
                }
            }
        } catch (IOException | InvalidConfigurationException e) {
            e.printStackTrace();
            return null;
        }

        return MappedYAML;
    }

    List<File> loadJarContent(String directory) {
        File folder = new File(plugin.getDataFolder(), directory);

        if (!folder.exists()) {
            folder.mkdirs();

            JarFile jarFile;
            try {
                jarFile = new JarFile(jar);
            } catch (IOException e) {
                e.printStackTrace();
                return null;
            }
            Enumeration<JarEntry> entries = jarFile.entries();

            while (entries.hasMoreElements()) {
                JarEntry entry = entries.nextElement();
                String name = entry.getName();

                if (!(name.startsWith(directory + "/")) || entry.isDirectory()) continue;

                if (!(new File(plugin.getDataFolder(), name).exists())) plugin.saveResource(name, false);
            }
        }

        return listFolderContent(new ArrayList<>(), folder);
    }

    List<File> listFolderContent(List<File> list, File folder) {
        for (File file : Objects.requireNonNull(folder.listFiles())) {
            if (file.isDirectory()) {
                listFolderContent(list, file);
            } else {
                list.add(file);
            }
        }
        return list;
    }
}

File: src/main/java/io/github/mg138/tsbook/setting/util/ConfigBuilder.kt
package io.github.mg138.tsbook.setting.util

import org.bukkit.configuration.ConfigurationSection
import org.bukkit.configuration.InvalidConfigurationException
import org.bukkit.configuration.file.YamlConfiguration
import org.bukkit.plugin.java.JavaPlugin
import java.io.File
import java.io.IOException
import java.util.*
import java.util.jar.JarFile
import java.util.function.Consumer

class ConfigBuilder(private val plugin: JavaPlugin, private val jar: File) {
    private val dataFolder = plugin.dataFolder
    private val jarFile = JarFile(jar)
    
    fun create(path: String, target: String): YamlConfiguration {
        val to = path + target
        val file = File(dataFolder, to)
        if (!file.exists()) {
            file.parentFile.mkdirs()
            plugin.saveResource(to, false)
        }
        return YamlConfiguration().also {
            it.load(file)
        }
    }

    fun createFolder(path: String, target: String): YamlConfiguration {
        val yaml = YamlConfiguration()
        loadJarContent(path).forEach { file ->
            if (file.name == target) {
                yaml.load(file)
            }
        }
        return yaml
    }

    fun createFolder(path: String): List<YamlConfiguration> {
        val list: MutableList<YamlConfiguration> = ArrayList()
        loadJarContent(path).forEach { file ->
            list.add(YamlConfiguration.loadConfiguration(file))
        }
        return list
    }

    fun createMap(path: String, key: String): Map<String, YamlConfiguration> {
        val map: MutableMap<String, YamlConfiguration> = HashMap()
        loadJarContent(path).forEach { file ->
            YamlConfiguration().apply {
                load(file)
                map[getString(key)!!] = this
            }
        }
        return map
    }

    fun createSectionMap(path: String): Map<String, ConfigurationSection> {
        val map: MutableMap<String, ConfigurationSection> = HashMap()
        loadJarContent(path).forEach { file ->
            YamlConfiguration().apply {
                load(file)
                getKeys(false).forEach { section ->
                    map[section] = getConfigurationSection(section)!!
                }
            }
        }
        return map
    }

    private fun loadJarContent(directory: String): List<File> {
        val folder = File( dataFolder, directory)
        if (!folder.exists()) {
            folder.mkdirs()

            val entries = jarFile.entries()
            while (entries.hasMoreElements()) {
                val entry = entries.nextElement()
                if (entry.isDirectory) continue

                val name = entry.name
                if (!name.startsWith("$directory/")) continue

                if (!File(dataFolder, name).exists()) plugin.saveResource(name, false)
            }
        }
        val result: MutableList<File> = ArrayList()
        recursiveListFolder(result, folder)
        return result
    }

    private fun recursiveListFolder(list: MutableList<File>, folder: File) {
        folder.listFiles()?.forEach { file ->
            if (file.isDirectory) {
                recursiveListFolder(list, file)
            } else {
                list.add(file)
            }
        }
    }
}